core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java

```diff
@@ -378,14 +389,57 @@ protected void stopServing() throws Exception {
         Thread.currentThread().interrupt();
       }
     }
-    if (mJvmPauseMonitor != null) {
-      mJvmPauseMonitor.stop();
+  }
+
+  protected void stopCommonServices(boolean force, boolean stopIfEnabled) throws Exception {
+    if (force || ServerConfiguration.getBoolean(
+        PropertyKey.STANDBY_MASTER_METRICS_SINK_ENABLED) == stopIfEnabled) {
+      MetricsSystem.stopSinks();
     }
+    if (force || ServerConfiguration.getBoolean(
+        PropertyKey.STANDBY_MASTER_WEB_ENABLED) == stopIfEnabled) {
+      stopServingWebServer();
+    }
+  }
+
+  /**
+   * Stops all services.
+   */
+  protected void stopServing() throws Exception {
+    stopLeaderServing();
+    stopCommonServices(true, true);
+    stopJvmMonitorProcess();
+  }
+
+  protected void stopServingWebServer() throws Exception {
     if (mWebServer != null) {
       mWebServer.stop();
       mWebServer = null;
     }
-    MetricsSystem.stopSinks();
+  }
+
+  protected void stopJvmMonitorProcess() {
+    if (mJvmPauseMonitor != null) {
+      mJvmPauseMonitor.stop();
+    }
+  }
+
+  @VisibleForTesting
+  public boolean waitForWebServerReady(int timeoutMs) {
```

Comment:
The existing `waitForReady()` should be renamed to `waitForGrpcServerReady()`. Otherwise it's really confusing, and you can already see the confusions in your unit test.

Reply:
`waitForReady` is a interface method of `Process`, so it's hard to rename it, but I can add a new method `waitForGrpcServerReady` to `MasterProcess`.